eventlog/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,293 @@
+
+func init() {
+	config.Register("eventer/cloudwatch", func(constr *config.Constructor) {
+		var sess *session.Session
+		constr.InstanceVar(&sess, "aws", "aws", "AWS configuration for all CloudWatch calls")
+		var group string
+		constr.StringVar(&group, "group", "eventlog", "the CloudWatch log group of the stream to which events will be sent")
+		var stream string
+		constr.StringVar(&stream, "stream", "", "the CloudWatch log stream to which events will be sent")
+		constr.Doc = "eventer/cloudwatch configures an eventer that sends events to a CloudWatch log stream"
+		constr.New = func() (interface{}, error) {
+			cw := cloudwatchlogs.New(sess)
+			if stream == "" {
+				// All the information of RFC 3339 but without colons, as stream
+				// names cannot have colons.
+				const layout = "20060102T150405.000-0700"
+				// The default stream name incorporates the current executable
+				// name and time for some measure of uniqueness and usefulness.
+				stream = strings.Join([]string{readExec(), time.Now().Format(layout)}, "~")
+			}
+			return NewCloudWatchEventer(cw, group, stream), nil
+		}
+	})
+}
+
+// CloudWatchEventer logs events to CloudWatch Logs.
+type CloudWatchEventer struct {
+	client cloudwatchlogsiface.CloudWatchLogsAPI
+	group  string
+	stream string
+
+	cancel func()
+
+	// eventc is used to send events that are batched and sent to CloudWatch
+	// Logs.
+	eventc chan event
+
+	// syncc is used to force syncing of events to CloudWatch Logs.
+	syncc chan struct{}
+	// syncDonec is used to block for syncing.
+	syncDonec chan struct{}
+
+	// donec is used to signal that the event processing loop is done.
+	donec chan struct{}
+
+	initOnce sync.Once
+	initErr  error
+
+	sequenceToken *string
+
+	// loggedFullBuffer prevents consecutive printing of "buffer full" log
+	// messages inside Event. We get a full buffer when we are overloaded with
+	// many messages. Logging each dropped message is very noisy, so we suppress
+	// consecutive logging.
+	loggedFullBuffer int32
+}
+
+type event struct {
+	timestamp  time.Time
+	typ        string
+	fieldPairs []interface{}
+}
+
+// NewCloudWatchEventer returns a *CloudWatchLogger. It does create the group or
+// stream until the first event is logged.
+func NewCloudWatchEventer(client cloudwatchlogsiface.CloudWatchLogsAPI, group, stream string) *CloudWatchEventer {
+	eventer := &CloudWatchEventer{
+		client:    client,
+		group:     group,
+		stream:    stream,
+		eventc:    make(chan event, 1024),
+		syncc:     make(chan struct{}),
+		syncDonec: make(chan struct{}),
+		donec:     make(chan struct{}),
+	}
+	var ctx context.Context
+	ctx, eventer.cancel = context.WithCancel(context.Background())
+	go eventer.loop(ctx)
+	return eventer
+}
+
+// Event implements Eventer.
+func (c *CloudWatchEventer) Event(typ string, fieldPairs ...interface{}) {
+	select {
+	case c.eventc <- event{timestamp: time.Now(), typ: typ, fieldPairs: fieldPairs}:
+		atomic.StoreInt32(&c.loggedFullBuffer, 0)
+	default:
```

Comment:
What's going on here?

Reply:
I attempted to explain this in the comment for `loggedFullBuffer`:
```
// loggedFullBuffer prevents consecutive printing of "buffer full" log
// messages inside Event. We get a full buffer when we are overloaded with
// many messages. Logging each dropped message is very noisy, so we suppress
// consecutive logging.
```

Does that make sense?